### ISSUE TITLE
Triton: run IOC on Triton control PC: implementation

### ISSUE THREAD

**Issue by Tom-Willemsen** [ISISComputingGroup/IBEX#3125]
As a member of the cryogenics team, I would like the Triton IOC to run on the Triton control PC rather than the instrument PC so that I can run diagnostics easily, without depending on which beamline the fridge is on.

#### Acceptance criteria
- As a user, I can control the Triton system when the IOC is running on the Triton control PC
- As a user, I can see any alarms that my IOC is generating in the alarms view
- As a user, I can create blocks pointing at the Triton IOC PVs
- As a member of the cryogenics team, I can look at archived data for Triton PVs (regardless of which beamline the fridge was on at the time)
- As a member of the cryogenics team, I can view and control PVs from the triton system
- As a member of the cryogenics team, I can configure which beamline the fridge is currently on
- As a developer, I can see IOC logs from the Triton system
- As a developer, I can start and stop the Triton IOC.

#### Notes
1. The purpose of this ticket is to implement the solution described in the comments of https://github.com/ISISComputingGroup/IBEX/issues/3091 for running the Triton IOC on the Triton's control PC

**Comment by DominicOram** [comment]
Rough picture of the proposed architecture: ![img_20180504_183611220](https://user-images.githubusercontent.com/10086797/39642747-956d2984-4fca-11e8-8d25-ed8f6d2d6a73.jpg) (should be covered in linked ticket but add a better drawn picture to the wiki as part of ticket)

Criteria of each group ![img_20180504_183618145](https://user-images.githubusercontent.com/10086797/39642773-a4808330-4fca-11e8-9914-69e50fc7679b.jpg)

**Comment by FreddieAkeroyd** [comment]
I guess configuring the IOC will be via globals.txt ?

**Comment by Tom-Willemsen** [comment]
Setting the instrument will be done via a PV, see comments of https://github.com/ISISComputingGroup/IBEX/issues/3091

For the triton there are no other configuration options to worry about (it will always connect to the triton on `localhost` with a fixed port number)

**Comment by KathrynBaker** [comment]
This solution is also of interest to the Cryogenics group (as part of a small project) so that they can monitor the systems centrally (not part of the HLM work).

**Comment by Tom-Willemsen** [comment]
Re-requested by Cryogenics group, so re-proposing.

They say that better remote diagnostics would help them to prevent somewhere around ~1 problem/cycle (where a "problem" has a lost beamtime impact of multiple hours and sometimes days).

**Comment by Tom-Willemsen** [comment]
After discussion with @John-Holt-Tessella regarding MySQL (instrument) archives specifically:
- We will run the central archive with a static logging config that always points at all of the remote IOCs
  * This will be manually updated when we add a new type of "remote" ioc. This is because we can't just use the standard archive config tool so when we want to do this automatically it will need a new tool writing. For this ticket (which just relates to Tritons) a static logging config + documenting how to add extra iocs in future will be sufficient
  * The PVs will be archived as `ME:TRITON1:...`, not `IN:LET:...`
- We won't run an instrument archive on the remote PCs
  * This might turn out to be required in future (there may be a need to access archived values from when the fridge was not on the network) but does not conflict with the archive above so could be added in a later ticket.
  * We don't yet have an adequate management strategy for how we would manage resources on a computer we can't see (if it's not on the network). This is a much wider discussion point, and should get it's own ticket if we decide/need to do it.
- We will not push a logging config from the remote IOC to the instrument PC.
  * If scientists want it logged, the answer has always been that they should put a block on it
  * If we decide to do this in future, we would need to write an extra tool that merges the logging config of the remote IOC and the rest of the instrument. The tool would also need to know how to translate `ME:TRITON:...` into the `IN:LET:...` aliases in it's logging config.

**Comment by FreddieAkeroyd** [comment]
How inportant is archiving the PVS? The central archive server is not critical for instrument operation and is also not on a system that is under our direct control.

**Comment by ChrisM-S** [comment]
If the central archive server is not related to the ISIS data archive (where all cite-able data from ISIS is archived) and is not something under our direct control in IBEX where/what on earth is it ?

**Comment by ChrisM-S** [comment]
If the central archive server is not related to the ISIS data archive (where all cite-able data from ISIS is archived) and is not something under our direct control in IBEX where/what on earth is it ?

**Comment by ChrisM-S** [comment]
If the central archive server is not related to the ISIS data archive (where all cite-able data from ISIS is archived) and is not something under our direct control in IBEX where/what on earth is it ?

**Comment by Tom-Willemsen** [comment]
@ChrisM-S this is a terminology difference. This is not related to the ISIS data archive in any way. This is the equivalent of the MySQL instrument archives for IOCs running on equipment PCs (such as tritons).

**Comment by John-Holt-Tessella** [comment]
We believe that it is not important to archive the PVs. They currently do not archive the PVs and so it cannot be critical. If they need to archive it hopefully they can plug it into the network. If they can not do this and really need to archive it then we can think about installing an archive locally on the instrument.  The reason we are keen to not have an archiving system running locally is:

- installing and running mySql on the triton PC seems like a risk, the PC itself has an unknown architecture
- once on we would need to monitor and support the mySql server running on the instrument. So it kept running and did not use up its disk space.

**Comment by ChrisM-S** [comment]
OK, that makes complete sense to me, just like an ITC standing alone in its rack unconnected doesn't record it's own data, even if it is running.  Once the equipment it is "plugged into" an instrument, it becomes part of someone's experiment managed by a scientist who decides what matters and designates those data items (by the block/PV mapping) for logging storage and preservation.

For the cryogenics team, the current R80 and soon to be R55 cryogenics instrument machines are proxy instruments so that if they do offline experiments, setups etc. these machines become the ones responsible for setting blocks that matter and preserving runs (if it is appropriate).  If they do some publishable science or investigation on a magnet (say), the route through to viewing analysing and archiving the data is already present (RB number instrument name & run number).

There is a useful distinction here which I'm only just realising and that is to identify things like the Tritons, beamline camera computers, possibly others as **"device" computers** rather than "instrument" computers.  Up until now they have generally been classified by me as "Ancillary control PCs" which is a bit vague as this has been an anything else category!  The main distinction between these and other PCs on the beam line is that they are involved in _control_ and _DAQ_, but are only there to facilitate access to some, usually more complex device

**Comment by Tom-Willemsen** [comment]
I am splitting out various tickets from this one, as they can be done in parallel once the framework is in place:
- https://github.com/ISISComputingGroup/IBEX/issues/4465
- https://github.com/ISISComputingGroup/IBEX/issues/4466
- https://github.com/ISISComputingGroup/IBEX/issues/4467
- https://github.com/ISISComputingGroup/IBEX/issues/4468

**Comment by Tom-Willemsen** [comment]
Pull requests:
- Main PR for RemoteIocServer: https://github.com/ISISComputingGroup/EPICS-inst_servers/pull/208
- Gateways: https://github.com/ISISComputingGroup/EPICS/pull/172
- Upgrade script for configuration schema: https://github.com/ISISComputingGroup/EPICS-upgrade/pull/29
- GUI: https://github.com/ISISComputingGroup/ibex_gui/pull/909

**Comment by FreddieAkeroyd** [comment]
Is this something we can "system test"?

**Comment by Tom-Willemsen** [comment]
Not easily within our existing test frameworks, because to system test this fully requires two co-operating machines and we don't currently have the infrastructure set up for this.

However I think it is adequately tested with:
- Unit tests for the new code in the RemoteIocServer
- Existing system tests of the blockserver (most of the functionality is reused from the blockserver)

I definitely think it would be possible to create the framework for testing accross multiple machines etc (maybe by spinning up a VM/container of some sort within the test?), but it is probably out of scope for this ticket which is already quite large.

**Comment by DominicOram** [comment]
@Tom-Willemsen do we need to update the `_base` component in the `master` branch too?

**Comment by Tom-Willemsen** [comment]
I think last time this was discussed we said that the `_base` component stays on the older version (with an appropriate `config_version.txt`) and then the upgrade script takes care of any changes from that version? If I remember rightly this was to try to reduce merge conflicts between master and instruments.

I'm happy to change it if I've misremembered though, let me know. Either way, we should document the decision on the wiki...

**Comment by John-Holt-Tessella** [comment]
I don't remember, either we should have written it down my bad. Do we run the upgrade script over a new config set by default? If not we should add it to base because other wise new configs won't get that change.

**Comment by DominicOram** [comment]
Is master regularly merged into instrument configs? It seems rather counter intuitive to me that master is actually incorrect and you need to run an additional script to correct it.

**Comment by Tom-Willemsen** [comment]
Yes, I think the deploy script automatically merges master into the instrument configs every time we deploy (see https://github.com/ISISComputingGroup/ibex_utils/blob/master/installation_and_upgrade/ibex_install_utils/install_tasks.py#L519 ). That might not be the best solution though.
